tango/integrations/wandb/step_cache.py

```diff
@@ -0,0 +1,227 @@
+        return wandb.Api(overrides={"entity": self.entity, "project": self.project})
+
+    def _acquire_step_lock_file(self, step: Union[Step, StepInfo], read_only_ok: bool = False):
+        return FileLock(
+            self.step_dir(step).with_suffix(".lock"), read_only_ok=read_only_ok
+        ).acquire_with_updates(desc=f"acquiring step cache lock for '{step.unique_id}'")
+
+    def _step_artifact_name(self, step: Union[Step, StepInfo]) -> str:
+        if isinstance(step, Step):
+            return step.name if step.name is not None else step.unique_id
+        else:
+            return step.step_name if step.step_name is not None else step.unique_id
+
+    def get_step_result_artifact(
+        self, step: Union[Step, StepInfo]
+    ) -> Optional[wandb.apis.public.Artifact]:
+        try:
+            return self.wandb_client.artifact(
+                f"{self.entity}/{self.project}/{self._step_artifact_name(step)}:{step.unique_id}",
+                type=ArtifactKind.STEP_RESULT.value,
+            )
+        except WandbError as exc:
+            if is_missing_artifact_error(exc):
+                return None
+            else:
+                raise
+
+    def create_step_result_artifact(
+        self, step: Step, objects_dir: Optional[PathOrStr] = None
+    ) -> None:
+        """
+        Create an artifact for the result of a step.
+        """
+        artifact = wandb.Artifact(
+            self._step_artifact_name(step), type=ArtifactKind.STEP_RESULT.value
+        )
+
+        # Add files
+        if objects_dir is not None:
+            artifact.add_dir(str(objects_dir))
+
+        # Log/persist the artifact to W&B.
+        artifact.save()
+        artifact.wait()
+
+        # Add an alias for the step's unique ID.
+        # Only after we've logged the artifact can we add an alias.
+        artifact.aliases.append(step.unique_id)
+        artifact.save()
+        artifact.wait()
+
+    def use_step_result_artifact(self, step: Union[Step, StepInfo]) -> None:
+        """
+        "Use" the artifact corresponding to the result of a step.
+        """
+        if wandb.run is None:
+            raise RuntimeError("This can only be called from within a W&B run")
+        wandb.run.use_artifact(
+            f"{self.entity}/{self.project}/{self._step_artifact_name(step)}:{step.unique_id}"
+        )
+
+    def __contains__(self, step: Any) -> bool:
+        if isinstance(step, (Step, StepInfo)):
+            cacheable = step.cache_results if isinstance(step, Step) else step.cacheable
+            if not cacheable:
+                return False
+
+            key = step.unique_id
+
+            # First check if we have a copy in memory.
+            if key in self.strong_cache:
+                return True
+            if key in self.weak_cache:
+                return True
+
+            # Then check if we have a copy on disk in our cache directory.
+            with self._acquire_step_lock_file(step, read_only_ok=True):
```

Comment:
Is there a possibility of the disk cache dir and W&B being out of sync?

Reply:
Not for an individual step, unless someone messes with the contents of the directory. But that shouldn't happen. W&B artifacts are versioned, and each local step cache directory is tied to a specified version of the artifact on W&B. A version of an artifact is immutable AFAIK.